firedrake/slate/slate.py

```diff
@@ -620,3 +607,21 @@ def __repr__(self):
     def _key(self):
         """Returns a key for hash and equality."""
         return (type(self), self.operands, self.actee)
+
+
+# Establishes levels of precedence for Slate tensors
+precedences = [
+    [Tensor],
+    [UnaryOp],
+    [Add, Sub],
+    [Mul, Action]
+]
+
+# Here we establish the precedence class attribute for a given
+# Slate tensor. Unary and binary operations will have higher
+# precedence over terminal Tensor objects. Similarly, tensor
+# action and multiplication will have higher precedence over
+# addition/subtraction
```

Comment:
If you want/need to assign precedence levels to them, shouldn't terminals appear to be the most strongly binding? That is, they never need parenthesis around them, because they "bind" more strongly anyway than their surrounding.

Also please be aware that Python3 is stricter when comparing different types. `1 < None` is fine in Python2, but is an error in Python3.